File: 13.Linked_List/javascript/10.PalindromeList.js
class Node{
    constructor(data){
        this.data = data
        this.next = null
    }
}

function buildLinkedList(arr){
    
    let head = new Node(arr[0])
    let current = head
    
    for(let i=1;i<arr.length;i++){
        current.next = new Node(arr[i])
        current = current.next
    }
    
    return head
}


class Solution{
    isPalindrome(head) {
        let values = []
        
        let current = head
        
        while(current !== null){
            values.push(current.data)
            current = current.next
        }
        
        let reverse = [...values].reverse()
        
        return values.every((val,indx)=>val===reverse[indx])
        
    }
}


function  PrintLinkedList(head){
    let res = ''
    let current = head
    
    while(current !==null){
        res += current.data + ' '
        current = current.next
    }
    
    console.log(res)
}


let arr = [1,2,1]
let Sol = new Solution()

let head = buildLinkedList(arr)
let isPalin = Sol.isPalindrome(head)
console.log(`is Palindrome: ${isPalin}`)
PrintLinkedList(head)

File: 13.Linked_List/python/10.PalindromeList.py
class ListNode:
    def __init__(self, data):
        self.data = data
        self.next = None

def isPalindrome(head):
    values = []
    
    # Traverse the linked list and collect node values
    current = head
    while current:
        values.append(current.data)
        current = current.next
    
    # Check if list of values is equal to its reverse
    return values == values[::-1]

def printLinkedList(head):
    current = head
    while current:
        print(current.data, end=" ")
        current = current.next
    print()

# 🔸 Hardcoded Palindrome List: 1 -> 2 -> 3 -> 2 -> 1
node1 = ListNode(1)
node2 = ListNode(2)
node3 = ListNode(3)
node4 = ListNode(2)
node5 = ListNode(1)

# Linking the nodes
node1.next = node2
node2.next = node3
node3.next = node4
node4.next = node5

head = node1

# Print list and check palindrome
printLinkedList(head)  # Output: 1 2 3 2 1
print("Is Palindrome?", isPalindrome(head))  # Output: True
